### PR TITLE
Python: major features for the Kernel Arguments, Function Result and new Prompt Templating

### LINKED ISSUE #4565: Python: Replace SKContext with Kernel and KernelArguments (KernelArguments is a replacement for ContextVariables)
In the .NET SDK, all of the method signatures were updated to take a Kernel object and KernelArguments object instead of SKContext. This PR is to both update all the method signatures and deprecate SKContext and Context variables.

### LINKED ISSUE #4631: Python: Add AI service selector to Python
In the .NET SDK, an AI Service Selector was added so that a prompt template could be resolved to the correct model (in case there are multiple).

This issue should track the following work:
- Introduce the AI Service Selector class
- Provide a default implementation of the AI Service Selector (replicate the same logic as the .NET version)
- Allow users to register custom service selectors on the kernel

### LINKED ISSUE #4856: Python: add ChatHistory concept in sync with dotnet and adapt chat_prompt_template accordingly

### LINKED ISSUE #5003: Python: Enable multiple-named arguments in function calls

### PR DESCRIPTION
**Pull request by eavanvalkenburg**
### Motivation and Context

Major work to replace the context with Kernel Arguments on the front-end and Function Results on the back.

Updated the function decorator to a new approach, in line with dotnet.

Revamps the way function are called, allowing native functions to be completely ignorant of SK, other then the decorator.

This also moves things into the new folder structure in sync with dotnet.

### Description

 
Adds:
- KernelArguments, a dict-like class that replaces the variables in KernelContext. Closes #4565
- FunctionResult, a class to hold the results of function executions, includes the function, the value and metadata, as well as two convenient function to get the value out of it (str and get_inner_content) the first is generic, the second specifically for KernelContent responses (from AI services).
- AI Service Selector class, has a default, which is based on the order in the arguments followed by the order in the functions, can be overridden to implement your own strategy. Closes #4631 
- Introduces ChatHistory and refactors the PromptTemplateConfig. Closes #4856, #4630
- Improves parsing of templates, will now all validate during creation and throw an error then, instead of some that do not check for validaty until used.
- Introduces named_args block and thereby the ability to have multiple arguments for a function call in a template. Closes #5003

Updates:
- kernel_function decorators, the parameter decorator was removed and instead we now use Annotated to add a description of a field and we get the type and required from the function definition.
- core plugins, use the new approach to kernel_function decorators.
- planners, template engines have all been updated to use the kernel and kernelarguments instead of Context.
- Events have been updated, now use kernelarguments and function_result
- Tokenizers support for named_args and improvements on parsing and checking.
- Kernel examples and notebooks to use the latest code.
- All unit and integration tests. There is more code coverage now than before.

Removed:
- kernelContext
- kernel_function_parameter_decorator
- delegate handling code for native functions
- file_io_plugin and tests
- SemanticFunctionConfig
- ChatPromptTemplate

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

**Review by moonbox3 — APPROVED** [review]
I've been reviewing all of the PRs up to this point and am approving all of these changes from the feature branch to main.

File: python/samples/kernel-syntax-examples/memory.py [hunk 1]
```diff
@@ -31,6 +33,7 @@ async def search_memory_examples(kernel: sk.Kernel) -> None:
         print(f"Answer: {result[0].text}\n")
 
 
+# TODO fix this ASAP
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/samples/kernel-syntax-examples/memory.py:36]: Yeah, but I think we'll fix it in main next week :)

File: python/notebooks/09-groundedness-checking.ipynb
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/notebooks/09-groundedness-checking.ipynb:1]: Thanks! Will implement on Monday.

File: python/notebooks/07-hugging-face-for-plugins.ipynb
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/notebooks/07-hugging-face-for-plugins.ipynb:1]: Good call -- will fix this in the next iteration on Monday.

**Review by juliomenendez — APPROVED** [review]

**Review by juliomenendez — APPROVED** [review]

**Review by juliomenendez — APPROVED** [review]

**Review by juliomenendez — APPROVED** [review]

**Pull request merged**